src/libraries/System.Private.CoreLib/src/System/Math.cs

```diff
@@ -1210,10 +1213,13 @@ public static double ReciprocalEstimate(double d)
         ///    <para>On hardware without specialized support, this may just return <c>1.0 / Sqrt(d)</c>.</para>
         /// </remarks>
         [Intrinsic]
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static double ReciprocalSqrtEstimate(double d)
         {
+#if MONO || TARGET_RISCV64 || TARGET_LOONGARCH64
```

Comment:
Is it intentional that this one has `TARGET_RISCV64 || TARGET_LOONGARCH64`, but `ReciprocalEstimate` does not? The JIT logic for both ReciprocalEstimate and ReciprocalSqrtEstimate look identical.

Reply:
Yes, the difference is that ReciprocalEstimate only requires trivial IL ops (1 / x) while ReciprocalSqrtEstimate requires support for the Sqrt intrinsic, which RV64 and LA64 don’t have yet. I didn’t think it was worth the complexity figuring out the plumbing to call COMDouble::Sqrt manually for these two cases and that when the community gets the support online we could just opt to fix those platforms up at that time.

I’ll log a tracking issue for them on it.

Reply:
Yes, this make sense.

Is there a matching ifdef or condition in the JIT where we do one thing for ReciprocalEstimate  and a different thing for ReciprocalSqrtEstimate on RiscV and oongArch?

Reply:
Not identical, but yes.

This comes down to the `IsTargetIntrinsic` handling which has a direct ifdef for RV64/LA64: https://github.com/dotnet/runtime/blob/1e0879f9ddcec7648d44414aa49833cf64cb540e/src/coreclr/jit/importercalls.cpp#L7478-L7481

The `IsMathIntrinsic` handling which is the same for all platforms: https://github.com/dotnet/runtime/blob/1e0879f9ddcec7648d44414aa49833cf64cb540e/src/coreclr/jit/importercalls.cpp#L7502

and the handling in `impEstimateIntrinsic` where there is an upfront `#ifdef` for platforms with `FEATURE_HW_INTRINSICS` support and then a path that handles the fallback for `ReciprocalEstimate`/`ReciprocalSqrtEstimate`: https://github.com/dotnet/runtime/blob/1e0879f9ddcec7648d44414aa49833cf64cb540e/src/coreclr/jit/importercalls.cpp#L8862-L8882

I can update the last bit to explicit assert that `NI_System_Math_ReciprocalSqrtEstimate` is unreached for `RV64`/`LA64`, but hadn't done it initially since the RV64 job was already asserting prior to the `#ifdef` in managed due to it being unhandled.

Reply:
I am not following what makes the last bit unreachable.

Let's assume that we are importing a `ReciprocalEstimate` or `ReciprocalSqrtEstimate` call with optimizations on. If I am reading the code correctly, this is going to be the flow in both cases:

1. `impIntrinsic` recognizes the intrinsic
2. `impIntrinsic` calls `impEstimateIntrinsic`
3. `impEstimateIntrinsic` skips `FEATURE_HW_INTRINSICS` (`FEATURE_HW_INTRINSICS` is not defined for RiscV/Loongarch)
4. `impEstimateIntrinsic` creates `1/op` for `ReciprocalEstimate` and `1/op^2` for `ReciprocalSqrtEstimate`. `1/op^2` is going to explode later since there is no intrinsic expansion for `NI_System_Math_Sqrt` on  RiscV/Loongarch as you have pointed out.

What am I missing?

Reply:
Ah, you're right. I broke this in between an early version of the PR and one of the last couple commits when I changed it from:
```csharp
#if MONO || ...
    public static float ReciprocalSqrtEstimate(float x) => 1.0f / Sqrt(x);
#else
    [Intrinsic]
    public static float ReciprocalSqrtEstimate(float x) => ReciprocalSqrtEstimate(x);
#endif 
```

into:
```csharp
    [Intrinsic]
    public static float ReciprocalSqrtEstimate(float x)
    {
#if MONO || ...
        return 1.0f / Sqrt(x);
#else
        return ReciprocalSqrtEstimate(x);
#endif 
    }
```

This fails for `impMathIntrinsic` because it has the check to `IsIntrinsicImplementedByUserCall(intrinsicName)` which itself calls `IsTargetIntrinsic`.

I need to ensure that `impEstimateIntrinsic` has the check also calls `IsIntrinsicImplementedByUserCall` and bails for cases that report `false`.

Reply:
Should be fixed now.